### PR TITLE
Implementing DyeableData

### PR DESCRIPTION
**Pull request by Rain336**
Implementation of `SpongeDyeableData`.
- [x] Registration of field getters and setters
- [x] Accurately used the appropriate `AbstractData` implementation

Implementation of `ImmutableSpongeDyeableData`
- [x] Accurately extended the appropriate `AbstractImmutableData` implementation
- [x] Accurately instantiating final instance fields with an `ImmutableValue` counter part for any value getters
- [x] If necessary, creating a new `ImmutableValue` for a value that should not be cached, or, using the existing caching utils.

Implementation of the `DyeColorValueProcessor`s:
- [x] Appropriately extended `AbstractSpongeValueProcessor`
- [x] Individual processors for each source type possible (one for `ItemStack`/`TileEntity`/`Entity` as necessary).

Registration:
- [x] Registered the `Key` correctly by `DYE_COLOR` in the `KeyRegistry`
- [x] Registered the `DataProcessor`s and `DataManipulatorBuilder` in `DataRegistrar`
- [x] Registered the `ValueProcessor`s in the `DataRegistrar`

Test Plugin code:
```java
@Plugin(id = "CodeTester", name = "CodeTester", version = "1.0.0")
public class CodeTester {

    @Listener
    public void preInit(GamePreInitializationEvent event) {
        CommandSpec getColor = CommandSpec.builder()
                .arguments(GenericArguments.catalogedElement(Texts.of("Color"), DyeColor.class), GenericArguments.string(Texts.of("Type")))
                .executor((src, args) -> {
                    DyeColor color = args.<DyeColor>getOne("Color").get();
                    String typeString = args.<String>getOne("Type").get();
                    CatalogType type;
                    if (src instanceof Player) {
                        Player player = (Player) src;
                        switch (typeString.toLowerCase()) {
                            case "glass_pane":
                                type = ItemTypes.STAINED_GLASS_PANE;
                                break;
                            case "glass":
                                type = ItemTypes.STAINED_GLASS;
                                break;
                            case "wool":
                                type = ItemTypes.WOOL;
                                break;
                            case "dye":
                                type = ItemTypes.DYE;
                                break;
                            case "sheep":
                                type = EntityTypes.SHEEP;
                                break;
                            case "wolf":
                                type = EntityTypes.WOLF;
                                break;
                            default:
                            case "clay":
                                type = ItemTypes.STAINED_HARDENED_CLAY;
                                break;
                        }
                        if (type instanceof ItemType) {
                            ItemStack item = ItemStack.of((ItemType) type, 1);
                            if (item.supports(Keys.DYE_COLOR)) {
                                item.offer(Keys.DYE_COLOR, color);
                            }
                            player.setItemInHand(item);
                            return CommandResult.success();
                        } else if (type instanceof EntityType) {
                            Vector3d position = player.getLocation().getPosition().add(2, 0, 2);
                            Optional<Entity> optional = player.getLocation().getExtent().createEntity((EntityType) type, position);
                            if (optional.isPresent()) {
                                Entity entity = optional.get();
                                if (entity.supports(Keys.DYE_COLOR)) {
                                    entity.offer(Keys.DYE_COLOR, color);
                                }
                                player.getLocation().getExtent().spawnEntity(entity, Cause.of(this));
                                return CommandResult.success();
                            }
                        }
                    }
                    return CommandResult.empty();
                }).build();
        Sponge.getCommandDispatcher().register(this, getColor, "getColor");
    }
}
```

### CODE REVIEW

**Comment by JBYoshi** [comment]
It's best if you use `DualProcessor`s whenever you can. They're a combination of the `DataProcessor` and the `ValueProcessor` for a `DataManipulator` that only has a single value (such as `DyeableData`).

**Comment by Rain336** [comment]
I know but there is currently only  a `AbstractSingleTargetDualProcessor` and this has three different targets

**Comment by JBYoshi** [comment]
Just create three different `DualProcessor`s and register one for each target.

**Comment by gabizou** [comment]
@Rain336 Do you have a chance to hop on our IRC channel, I'd like to speak to you about something.

**Comment by JBYoshi** [comment]
@Rain336 Next time, make sure to post in #8 to let everyone know you're working on a particular manipulator. I've made a post for `DyeableData` for you.

File: src/main/java/org/spongepowered/common/data/processor/value/ItemDyeColorValueProcessor.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import org.spongepowered.api.data.type.DyeColors;
+import org.spongepowered.api.data.value.ValueContainer;
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.data.value.mutable.Value;
+import org.spongepowered.common.data.processor.common.AbstractSpongeValueProcessor;
+import org.spongepowered.common.data.value.SpongeValueFactory;
+
+import java.util.Optional;
+
+public class ItemDyeColorValueProcessor extends AbstractSpongeValueProcessor<ItemStack, DyeColor, Value<DyeColor>> {
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/processor/value/ItemDyeColorValueProcessor.java]: This needs to check if the `ItemStack` is supported.

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/SpongeDyeableData.java [hunk 2]
```diff
@@ -36,7 +34,7 @@
 public class SpongeDyeableData extends AbstractSingleCatalogData<DyeColor, DyeableData, ImmutableDyeableData> implements DyeableData {
 
     public SpongeDyeableData(DyeColor dyeColor) {
-        super(DyeableData.class, checkNotNull(dyeColor), Keys.DYE_COLOR, ImmutableDyeableData.class);
+        super(DyeableData.class, dyeColor, Keys.DYE_COLOR, ImmutableDyeableData.class);
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/SpongeDyeableData.java]: Why did you remove this?

Comment (Rain336) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/SpongeDyeableData.java]: The Manipulator Test failed because of this

Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/SpongeDyeableData.java]: You'll want to add `@ImplementationRequiredForTest` to the class, which prevents the ManipulatorTest from running (since the dye colors aren't available in the testing environment).

**Comment by Aaron1011** [comment]
@Rain336: The values seem incorrect for the `dye` item - for example, using `black` with your test plugin gives me bone meal.

File: src/main/java/org/spongepowered/common/data/processor/value/ItemDyeColorValueProcessor.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    @Override
+    protected boolean set(ItemStack container, DyeColor value) {
+        container.setItemDamage(((EnumDyeColor) (Object) value).getMetadata());
+        return true;
+    }
+
+    @Override
+    protected Optional<DyeColor> getVal(ItemStack container) {
+        return Optional.of((DyeColor) (Object) EnumDyeColor.byMetadata(container.getItemDamage()));
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/processor/value/ItemDyeColorValueProcessor.java]: Use `EnumDyeColor.byDyeDamage(stack.getMetadata())` - that's what `ItemDye` uses. This should fix what @Aaron1011 mentioned

**Comment by Aaron1011** [comment]
@Rain336: Can you change `SpongeDyeableData` to properly pass `ImmutableSpongeDyeableData` to the superclass constrcutor? Right now, it's passing `ImmutableDyeableData`

**Comment by Aaron1011** [comment]
@Rain336: Now dyes work, but all of the other items (wool, glass, etc) are broken :P

**Comment by gabizou** [comment]
@Rain336 Could you squash or rebase your commits so that there's one clean commit instead of multiple merges and small fixes commits?
